File: c/Camera.c
/*
 * Camera.c
 */

#include "Camera.h"
#include "Image.h"
#include <math.h>
#include <Accelerate/Accelerate.h>

Camera cam;

void CameraCalcViewDir()
{
	V3Subtract(&(cam.dir),cam.lookAt,cam.eye);
	V3Normalize(&(cam.dir));
}

void CameraInit(int width, int height)
{
	float fov;
	cam.b.z = 0.0001f;
	cam.t.z = cam.b.z + 1.0f;
	fov = M_PI * (cam.fov / 360.0f);
	//tan(fov) = t.y / b.z
	cam.t.y = cam.b.z * tanf(fov);
	//t.x / t.y = nx / ny
	cam.t.x = ((float)width / (float)height) * cam.t.y;
	cam.b.x = -(cam.t.x);
	cam.b.y = -(cam.t.y);
	CameraCalcViewDir();
	CameraUpdate();
}

void CameraUpdate()
{
	cam.w = cam.dir;

	V3Cross(&(cam.u), cam.up, cam.w);
	V3Normalize(&(cam.u));

	V3Cross(&(cam.v), cam.w, cam.u);
}

void CameraEyeRay(Ray *r, int x, int y, int width, int height, int subx, int suby, int antiAliasKernelSize)
{
	Vector3 result,uprime,vprime,wprime;
	// x-flip ?
	x = width - x;
	// y-flip ?
//	y = height - y;
	result.x = cam.b.x + ((cam.t.x - cam.b.x) * ((float)x + ((float)(subx+1)/(antiAliasKernelSize+1))) / (float)width);
	result.y = cam.b.y + ((cam.t.y - cam.b.y) * ((float)y + ((float)(suby+1)/(antiAliasKernelSize+1))) / (float)height);
	result.z = cam.b.z;
	uprime = cam.u;
	cblas_sscal(3, result.x, (float*)&uprime, 1);
	vprime = cam.v;
	cblas_sscal(3, result.y, (float*)&vprime, 1);
	wprime = cam.w;
	cblas_sscal(3, result.z, (float*)&wprime, 1);
	//with respect to the eye
	result.x = uprime.x + vprime.x + wprime.x;
	result.y = uprime.y + vprime.y + wprime.y;
	result.z = uprime.z + vprime.z + wprime.z;
	V3Normalize(&result);
    r->o = cam.eye;
	r->d = result;
}

void CameraForward(float dist)
{
	Vector3 toMove = cam.dir;

	cblas_sscal(3, dist, (float*)&toMove, 1);
	V3AddTo(&(cam.eye), toMove);
	CameraUpdate();
}

void CameraBack(float dist)
{
	Vector3 toMove = cam.dir;

	cblas_sscal(3, -dist, (float*)&toMove, 1);
	V3AddTo(&(cam.eye), toMove);
	CameraUpdate();
}

void CameraLeft(float dist)
{
	Vector3 toMove;

	V3Cross(&toMove, cam.up, cam.dir);
	cblas_sscal(3, dist, (float*)&toMove, 1);
	V3AddTo(&(cam.eye), toMove);
	CameraUpdate();
}

void CameraRight(float dist)
{
	Vector3 toMove;

	V3Cross(&toMove, cam.up, cam.dir);
	cblas_sscal(3, -dist, (float*)&toMove, 1);
	V3AddTo(&(cam.eye), toMove);
	CameraUpdate();
}

File: rust/src/camera.rs
use vector::Vector3;
use ray::Ray;
use std::f64;

#[derive(Debug)]
pub struct Camera {
    width: f64,
    height: f64,

    pub eye: Vector3<f64>,
    pub up: Vector3<f64>,
    pub dir: Vector3<f64>,
    pub lookat: Vector3<f64>,
    pub fov: f64,
    u: Vector3<f64>,
    v: Vector3<f64>,
    w: Vector3<f64>,
    b: Vector3<f64>,
    t: Vector3<f64>,
}

impl Camera {
    pub fn new(width: u32, height: u32) -> Camera {
        let w = width as f64;
        let h = height as f64;
        let mut result = Camera {
            width: w,
            height: h,

            eye: Vector3::new(0.0, 0.0, 1.0),
            lookat: Vector3::init(0.0),
            up: Vector3::new(0.0, 1.0, 1.0),
            dir: Vector3::new(0.0, 0.0, -1.0),
            fov: f64::consts::PI / 4.0,

            // u, v, w are set by update()
            u: Vector3::new(0.0, 0.0, 0.0),
            v: Vector3::new(0.0, 0.0, 0.0),
            w: Vector3::new(0.0, 0.0, 0.0),

            // b and t are set by resize()
            b: Vector3::new(0.0, 0.0, 0.0),
            t: Vector3::new(0.0, 0.0, 0.0),
        };
        result.resize(w, h);
        result.update();
        result
    }

    pub fn resize(&mut self, width: f64, height: f64) {
        let bz = 0.0001;
        let tz = bz + 1.0;
        let fov = f64::consts::PI / 4.0;
        //tan(fov) = t.y / b.z
        let ty = bz * fov.tan();
        //t.x / t.y = nx / ny
        let tx = ((width as f64) / (height as f64)) * ty;
        let bx = -(tx);
        let by = -(ty);
        self.b = Vector3 { x: bx, y: by, z: bz };
        self.t = Vector3 { x: tx, y: ty, z: tz };
        self.width = width;
        self.height = height;
    }

    pub fn update(&mut self) {
        self.dir = (self.lookat - self.eye).normalize();
        self.w = self.dir;
        self.u = self.up.cross(self.w).normalize();
        self.v = self.w.cross(self.u).normalize();
    }

    pub fn eye_ray(&self, x: u32, y: u32) -> Ray {
        // x-flip ?
        let x = self.width as u32 - x;
        // y-flip ?
        let y = self.height as u32 - y;
        let nx = ((x as f64) + 0.5) / self.width;
        let ny = ((y as f64) + 0.5) / self.height;
        let sx = self.b.x + (self.t.x - self.b.x) * nx;
        let sy = self.b.y + (self.t.y - self.b.y) * ny;
        let sz = self.b.z;
        let uprime = self.u * sx;
        let vprime = self.v * sy;
        let wprime = self.w * sz;
        //with respect to the eye
        let rd = Vector3 {
            x: uprime.x + vprime.x + wprime.x,
            y: uprime.y + vprime.y + wprime.y,
            z: uprime.z + vprime.z + wprime.z,
        };
        Ray {
            origin: self.eye,
            direction: rd.normalize(),
        }
    }
}
